### ISSUE TITLE
Attach self._patch to images in adversarial_patch_pytorch.py

### ISSUE THREAD

**Issue by shuoli90** [Trusted-AI/adversarial-robustness-toolbox#1329]
**Describe the bug**

First, thank you for providing this package!
I read through the training steps in adversarial_patch_pytorch.py. It seems that self._patch needs to be attached to original images before computing the loss in self._train_step function. Otherwise, after computing the loss, we will not get the gradients with respect to the patch. 

**Expected behavior**
Add _random_overlay before calling self._loss.

**Screenshots**
![Screen Shot 2021-09-29 at 8 56 08 AM](https://user-images.githubusercontent.com/19268975/135272982-42c0309b-caa4-4828-b093-55f73ff95e8d.png)

**Comment by beat-buesser** [comment]
Hi @shuoli90 Thank you very much for using ART and your thanks!

I think the patch should be added to the images in method `_random_overlay` in https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/49e13eac6f4d1154ccfdc5eb398a93924455fec6/art/attacks/evasion/adversarial_patch/adversarial_patch_pytorch.py#L409
and https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/49e13eac6f4d1154ccfdc5eb398a93924455fec6/art/attacks/evasion/adversarial_patch/adversarial_patch_pytorch.py#L180
is making sure that the optimisation accounts for the clip values.

Did you observe that the patch does not get trained or added to the images?

**Comment by shuoli90** [comment]
- I agree that _random_overlay function attaches patches to images. But there is no call to _random_overley before computing the loss in the training process. In other words, you will compute gradients on the original image instead of patched images.
- In your current implementation, it seems that you will not update self._patch inside _train_step. So, the call to torch.clamp will make no difference.
-  I just started reading the implementation and found that the current implementation may not find adversarial patches.

Would you mind taking a look at it?

**Comment by beat-buesser** [comment]
Ok. I think it is called the following way: variable `loss` in method `_loss` is based the output of method `_predictions` which calls `_random_overlay`.

**Comment by shuoli90** [comment]
I see. Thanks a lot for your explanation!
